src/rgw/rgw_amqp.cc

```diff
@@ -0,0 +1,240 @@
+          case AMQP_CHANNEL_CLOSE_METHOD:
+            ss << "server channel error: ";
+            break;
+          default:
+            ss << "server unknown error: ";
+            break;
+        }
+        amqp_connection_close_t* m = (amqp_connection_close_t*)reply.reply.decoded;
+        ss << m->reply_code << " text: " << std::string((char*)m->reply_text.bytes, m->reply_text.len);
+        return ss.str();
+      }
+    default:
+      ss << "unknown error, method id: " << reply.reply.id;
+      return ss.str();
+  }
+}
+
+// amqp connection guard
+// could be used for automatic cleanup when getting out of scope
+class ConnectionGuard {
+  private:
+    amqp_connection_state_t conn;
+  public:
+    ConnectionGuard(amqp_connection_state_t _conn) : conn(_conn) {}
+    ~ConnectionGuard() {
+      if (conn) {
+        amqp_destroy_connection(conn);
+      }
+    }
+    // call disable if cleanup is not needed anymore
+    void reset() {
+      conn = nullptr;
+    }
+};
+
+// TODO: why would you need multiple channels?
+static const amqp_channel_t CHANNEL_ID = 1;
+
+// utility function to create a connection
+amqp_connection_state_t create_connection(const amqp_connection_info& info, const std::string& exchange) {
+  // create connection
+  auto conn = amqp_new_connection();
+  ConnectionGuard guard(conn);
+  if (!conn) {
+    // failed to allocate connection
+    return nullptr;
+  }
+
+  // create and open socket
+  auto socket = amqp_tcp_socket_new(conn);
+  if (!socket) {
+    // failed to allocate socket
+    return nullptr;
+  }
+  auto rc = amqp_socket_open(socket, info.host, info.port);
+  if (rc < 0) {
+    // failed to open socket to broker
+    // TODO: retry
+    return nullptr;
+  }
+
+  // login to broker
+  auto reply = amqp_login(conn,
+      info.vhost, 
+      AMQP_DEFAULT_MAX_CHANNELS,
+      AMQP_DEFAULT_FRAME_SIZE,
+      0, // no heartbeat TODO: add conf
+      AMQP_SASL_METHOD_PLAIN, // TODO: add other types of security
+      info.user,
+      info.password);
+  if (reply.reply_type != AMQP_RESPONSE_NORMAL) {
+    // TODO: report the specific error and retry
+    return nullptr;
+  }
+
+  // open channel
+  amqp_channel_open(conn, CHANNEL_ID);
+  if (amqp_get_rpc_reply(conn).reply_type != AMQP_RESPONSE_NORMAL) {
+    // TODO: report the specific error and retry
+    return nullptr;
+  }
+
+  // verify that the topic exchange is there
+  amqp_exchange_declare(conn, 
+      CHANNEL_ID,
+      amqp_cstring_bytes(exchange.c_str()),
+      amqp_cstring_bytes("topic"),
+      0, // TODO: should be passive - exchange must already exist on broker
+      1, // durable
+      0, // dont auto-delete
+      0, // not internal
+      amqp_empty_table);
+
+  if (amqp_get_rpc_reply(conn).reply_type != AMQP_RESPONSE_NORMAL) {
+    // TODO: report the specific error and retry
+    return nullptr;
+  }
+
+  guard.reset();
+  return conn;
+}
+
+typedef std::map<amqp_connection_info, amqp_connection_state_t> ConnectionList;
```

Comment:
in principle, we try to avoid using std::list;  if the main type here is declared in an amqp header, perhaps we should wrap it.  ideally, I'd use boost::intrusive, but std::vector otherwise

Reply:
also, below, we are searching the list--maybe that suggests boost::intrusive avl_set

Reply:
looks like this is an older version. anyway, I'm using an std::unordered_map (should probably rename the type to ConnectionMap...).
for large number of connections this would be ideal, as the search is fast (just the hashing).
if we want to optimize for the case where the number of connections is small, I would change to an std::vector. I use the container to actually store the connections